app/domain/authentication/authn_jenkins/authenticator.rb

```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+require 'json'
+module Authentication
+  module AuthnJenkins
+    class Authenticator
+
+      Err = Errors::Authentication::AuthnJenkins
+      def initialize(env:)
+        @env = env
+      end
+
+      def webservice()
+        @webservice ||= ::Authentication::Webservice.new(
+          account:            @account,
+          authenticator_name: @authenticator_name,
+          service_id:         @service_id
+        )
+      end
+
+      def http_get_request(url, username, password) 
+        begin
+          uri = URI.parse(url)
+          http = Net::HTTP.new(uri.host, uri.port)
+          http.read_timeout = 5
+          request = Net::HTTP::Get.new(uri.request_uri)
+          request.basic_auth(username, password)
+          response = http.request(request)
+        rescue Net::OpenTimeout => e
+          Rails.logger.error("Timeout to Jenkins host Exception #{e.to_s}")
+          Rails.logger.error("Jenkins host '#{url}' is most likely invalid. Validate this url can be accessed from the conjur server.")
+          raise Err::HostNotFound, url
+        end
+
+        return response
+      end
+
+      def get_build_info(jenkinsURL, jobPath, buildNumber, username, password)
+        jenkinsBuildEndpoint = "/job/#{jobPath}/#{buildNumber}/api/json"
+        jenkinsBuildURL = "#{jenkinsURL}#{jenkinsBuildEndpoint}"
+        response = http_get_request(jenkinsBuildURL, username, password)
+        return response
+      end
+
+      def build_running?(response)
+        Rails.logger.debug("Jenkins Build Response: #{response.body}")
+
+        if response.code != '200'
+          raise Err::BuildInfoError.new(jenkinsBuildURL, "#{response.code} - #{response.body}")
+        end
+
+        json = JSON.parse(response.body)
+        return json['building']
+      end
+
+      def jenkins_public_key(response)
+        publicKey = response['X-Instance-Identity']
+
+        publicKey = "-----BEGIN PUBLIC KEY-----\n#{publicKey}\n-----END PUBLIC KEY-----"
+        Rails.logger.debug("Returned public key: #{publicKey}")
+        return OpenSSL::PKey::RSA.new(publicKey)
+      end
+
+      def parse_metadata(username, password)
+        jsonBody = JSON.parse password
+
+        buildNumber = jsonBody["buildNumber"]
+        signature = Base64.decode64(jsonBody["signature"])
+        jobProperty_hostPrefix = jsonBody["jobProperty_hostPrefix"]
+        if jsonBody.key?("jobProperty_hostPrefix")
+          jobName = username.sub("host/#{jobProperty_hostPrefix}/", "")
+        else
+          jobName = username.sub("host/", "")
+        end
+        jobPath = jobName.split("/").join("/job/")
+
+        return jobName, jobPath, buildNumber, signature
+      end
+
+      def valid?(input)
+        @account = input.account
```

Comment:
Try to avoid setting class level variables like this. It causes a couple of challenges. First, it means the `webservice` method interacts with this method (`valid?` in a non-intuitive way.  It also makes it harder to test the `webservice` method because class level variables need to be set outside the method being tested.

A possibility for resolving this could be to instantiate the webservice object in the method:
```ruby
def webservice(account, authenticator_name, service_id)
  @webservice ||= ::Authentication::Webservice.new(
    account:            account,
    authenticator_name: authenticator_name,
    service_id:         service_id
  )
end
...
def valid?(input)
  variables = webservice(
    input.account,
    input.authenticator_name,
    input.service_id
  )

  # Get needed secrets to connect into the jenkins API
  jenkinsUsername = variables.variable("jenkinsUsername").secret.value
...
```

Reply:
That makes sense. Resolved in this commit: https://github.com/AndrewCopeland/conjur/commit/f3972bb8aa834dc3c0037ec010425a6b916680e7